Corner orientation

lispandfound opened the pull request:
This pull request includes changes to the `source_modelling/sources.py` file, focusing on refactoring and improving the robustness of the code, as well as adding new tests in the `tests/test_sources.py` file to ensure the correctness of the changes.

### Refactoring and Improvements:

* Changed `@staticmethod` to `@classmethod` for factory methods in `Point`, `Plane`, and `Fault` classes, and replaced explicit class names with `Self` for better maintainability and readability. [[1]](diffhunk://#diff-5f2630533f745c6079637bcfdab384f3d08c41a5b14c36173b4dbc55423dc984L58-R60) [[2]](diffhunk://#diff-5f2630533f745c6079637bcfdab384f3d08c41a5b14c36173b4dbc55423dc984L75-R76) [[3]](diffhunk://#diff-5f2630533f745c6079637bcfdab384f3d08c41a5b14c36173b4dbc55423dc984L183-R208) [[4]](diffhunk://#diff-5f2630533f745c6079637bcfdab384f3d08c41a5b14c36173b4dbc55423dc984L200-R221) [[5]](diffhunk://#diff-5f2630533f745c6079637bcfdab384f3d08c41a5b14c36173b4dbc55423dc984L334-R356) [[6]](diffhunk://#diff-5f2630533f745c6079637bcfdab384f3d08c41a5b14c36173b4dbc55423dc984L445-R467) [[7]](diffhunk://#diff-5f2630533f745c6079637bcfdab384f3d08c41a5b14c36173b4dbc55423dc984L459-R480)
* Modified the `Plane` class constructor to validate the input bounds and ensure they form a valid plane.
* Updated `length_m` and `width_m` properties in the `Plane` class to explicitly return `float` values.
* Added `numpy.typing` as `npt` for type hinting in `Plane` class constructor.

### Testing Enhancements:

* Added new tests in `tests/test_sources.py` to validate the `Plane` class constructor with various invalid inputs, ensuring that the constructor raises appropriate errors.

claudio525 commented on source_modelling/sources.py:
```diff
@@ -442,8 +474,103 @@ class Fault:
+        if len(self.planes) == 1:
+            return
+
+        # Planes can only have one dip, dip direction, and width.
```
This holds for all NSHM faults?

lispandfound replied:
It does for everything except the dip, which I should actually remove because the dip direction and width are the only things that really needs to be constant.

lispandfound commented on source_modelling/sources.py:
```diff
@@ -180,11 +184,39 @@ class Plane:
-    def from_corners(corners: np.ndarray) -> "Plane":
+    bounds: npt.NDArray[np.float32]
+
+    def __post_init__(self):
```
Done and done I think.

lispandfound commented on source_modelling/sources.py:
```diff
@@ -282,16 +314,17 @@ def geometry(self) -> shapely.Polygon:
         length: float,
         projected_width: float,
-    ) -> "Plane":
+    ) -> Self:
```
Yeah this is really nice, always hated the string types.

claudio525 commented on source_modelling/sources.py:
```diff
@@ -569,6 +647,16 @@ def __post_init__(self) -> None:
         ]
 
+    @property
+    def dip(self) -> float:
```
If this can vary per plane, not sure if just returning the first one is the correct approach? Maybe the average (and rename to `avg_dip`)?

lispandfound replied:
Dip actually cannot vary per plane of a Fault. Supposing you did have two planes with a different dip, then they would not properly meet at an edge. It'd look like the picture below.

![image](https://github.com/user-attachments/assets/80436364-5cf4-420c-8911-3744ab1b58f1)

lispandfound replied:
Accordingly, I have added that consistency check back into the sources module.

lispandfound commented:
Tests fail here, but that's actually because the tests are not generating good planes!

claudio525 commented on source_modelling/sources.py:
```diff
@@ -217,8 +220,73 @@ class Plane:
+        # bottom vectors. If the dot product is positive then they are pointing
+        # in roughly the same direction, if it is negative then they are
+        # pointing in opposite directions.
+        if np.dot(top[1] - top[0], bottom[1] - bottom[0]) < 0:
```
```suggestion
        if np.dot(top[1] - top[0], bottom[0] - bottom[1]) < 0:
```
As bottom points are defined in opposite order to top points for the right-hand rule definition.

claudio525 replied:
Actually, correction, this doesn't produce incorrect results as the order of bottom points is switched in line 289. I do think it would be more logical to make this change, and change line 289 to be 
```
self.bounds = np.array([top[0], top[1], bottom[0], bottom[1]])
```
that way correct (i.e. following right hand rule) corner definition is not swapped.

lispandfound replied:
Yeah I was wondering why my tests weren't picking this "bug" up and then saw your comment.

claudio525 commented on source_modelling/sources.py:
```diff
@@ -217,8 +220,73 @@ class Plane:
+        if not np.isclose(orientation, 0) and orientation < 0:
+            top = top[::-1]
+            bottom = bottom[::-1]
+        self.bounds = np.array([top[0], top[1], bottom[1], bottom[0]])
```
As per comment above
```suggestion
        self.bounds = np.array([top[0], top[1], bottom[0], bottom[1]])
```

lispandfound commented:
Since the last review:

1. I've added test cases to check that the Fault class actually rejects invalid faults.
2. Fixed the tests that were failing because of API changes introduced by #16.
3. Misc ruff fixes.

joelridden commented on source_modelling/sources.py:
```diff
@@ -809,7 +810,7 @@ def __post_init__(self) -> None:
         """
         # The trivial case where the number of planes is one should be ignored
-        if len(self.planes) == 1:
+        if len(self.planes) <= 1:
```
Are they ever 0?

lispandfound replied:
sometimes, when the fault is constructed dynamically in the case of the NSHMDB
